Windows Support

munyirik opened the pull request:
I started working on Windows support for i2c-bus and this is my initial PR. The need for this came about when I started working on a project that requires i2c and runs on [Windows IoT Core](https://developer.microsoft.com/en-us/windows/iot).
In the implementation, what I did was try to match the i2c-bus functions with the methods exposed by the [i2cDevice](https://msdn.microsoft.com/library/windows/apps/windows.devices.i2c.i2cdevice.aspx) class on Windows.

There are still functions that are not implemented yet but I wanted to get this out early to get feedback as soon as possible. Note, there might be some functions that stay 'not implemented' since I can only use what i2cDevice exposes. It does however provide basic functionality to do basic read/write which was sufficient for my project. I also added an example on how to use it to read temperature from a [MCP9808](https://learn.adafruit.com/adafruit-mcp9808-precision-i2c-temperature-sensor-guide/overview) temperature sensor.

Building it does require a [change I made to gyp](https://chromium.googlesource.com/external/gyp/+/02b145a1a4f4e1c62e8bae06045caf852d9ef17f) that is not yet in the node branch. So until that change is merged to the node Github repo, building will require the following changes to gyp in your node installation:
1.Replace \<Node.js install path>\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py with [msvs.py.txt](https://github.com/fivdi/i2c-bus/files/262154/msvs.py.txt)
2.Replace \<Node.js install path>\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\MSVSSettings.py with [MSVSSettings.py.txt](https://github.com/fivdi/i2c-bus/files/262155/MSVSSettings.py.txt)
[Edit] **be sure to remove the .txt extensions from the files above**

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+      "(string i2cControllerName)"));
+  }
+
+  Nan::HandleScope scope;
```
An implicit HandleScope is created on JavaScript accessible methods. Why is another one created here?

munyirik replied:
Ah yeah, I'll take it out. And for the rest of the methods as well...

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+      "(int length, Buffer buffer)"));
+  }
+
+  Nan::HandleScope scope;
```
An implicit HandleScope is created on JavaScript accessible methods. Why is another one created here?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+      "incorrect arguments passed to WinI2c::Write"));
+  }
+
+  Nan::HandleScope scope;
```
An implicit HandleScope is created on JavaScript accessible methods. Why is another one created here?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+      "(int length, Buffer buffer)"));
+  }
+  
+  Nan::HandleScope scope;
```
An implicit HandleScope is created on JavaScript accessible methods. Why is another one created here?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+      "(int cmd, int length, Buffer buffer)"));
+  }
+
+  Nan::HandleScope scope;
```
An implicit HandleScope is created on JavaScript accessible methods. Why is another one created here?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+  DeviceInformationCollection^ i2cDeviceControllers = create_task(DeviceInformation::FindAllAsync(deviceSelector)).get();
+
+  if (nullptr == i2cDeviceControllers) {
+    Nan::ThrowError(Nan::ErrnoException(EPERM, "WinI2c::OpenSync", "DeviceInformation::FindAllAsync failed to return controller(s)"));
```
Should this not be return Nan::ThrowError(...)?

munyirik replied:
Will fix...

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+    
+    obj->_i2cDevice = create_task(I2cDevice::FromIdAsync(obj->_i2cDeviceId, settings)).get();
+    if (nullptr == obj->_i2cDevice) {
+      Nan::ThrowError(Nan::ErrnoException(EPERM, "WinI2c::SetDevice", "I2cDevice::FromIdAsync failed to return an i2c device"));
```
Although not necessary, would return Nan::ThrowError(...) be better here?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+    }
+  }
+  catch (Exception^ e) {
+    Nan::ThrowError(Nan::ErrnoException(e->HResult, "WinI2c::ReadPartial", ""));
```
Should this not be return Nan::ThrowError(...)?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+  try {
+    obj->_i2cDevice->Write(ArrayReference<BYTE>(bytes.data(), static_cast<unsigned int>(bytes.size())));
+  } catch (Exception^ e) {
+    Nan::ThrowError(Nan::ErrnoException(e->HResult, "WinI2c::Write", ""));
```
Although not necessary, would return Nan::ThrowError(...) be better here?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+    }
+  }
+  catch (Exception^ e) {
+    Nan::ThrowError(Nan::ErrnoException(e->HResult, "WinI2c::WritePartial", ""));
```
Should this not be return Nan::ThrowError(...)?

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,318 @@
+    }
+  }
+  catch (Exception^ e) {
+    Nan::ThrowError(Nan::ErrnoException(e->HResult, "WinI2c::WriteReadPartial", ""));
```
Should this not be return Nan::ThrowError(...)?

fivdi commented:
>I started working on Windows support for i2c-bus and this is my initial PR.

Windows support for i2c-bus would be really good.

>The need for this came about when I started working on a project that requires i2c and runs on Windows IoT Core.

I hope it's not an impolite question but what project are you working on? Is it an OSS project?

>There are still functions that are not implemented yet but I wanted to get this out early to get feedback as soon as possible.

I took a quick look at the PR and added some comments. I'll take a longer look soon. So far some of the sync functions have been implemented. Will async functions also be implemented?

>Note, there might be some functions that stay 'not implemented' since I can only use what i2cDevice exposes.

Do you have an idea what functions might stay not implemented?

>Building it does require a change I made to gyp that is not yet in the node branch. So until that change is merged to the node Github repo, building will require the following changes to gyp in your node installation:
>1.Replace <Node.js install path>\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py with msvs.py.txt
>2.Replace <Node.js install path>\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\MSVSSettings.py with MSVSSettings.py.txt

I don't have a Windows machine to build it. I'll see if I can arrange one.

munyirik commented:
> Windows support for i2c-bus would be really good.

Awesome!

>I hope it's not an impolite question but what project are you working on? Is it an OSS project?

Not impolite at all:-) !. Yes it's OSS. I'm working on getting [this](https://www.youtube.com/watch?v=4lSukCXqQV4) hexapod working on Windows with Node.js.

> I took a quick look at the PR and added some comments. I'll take a longer look soon. So far some of the sync functions have been implemented. Will async functions also be implemented?

Yes. I'm planning to implement those as well. 

> Do you have an idea what functions might stay not implemented?

An example might be i2cFuncs. But I'll do some digging to see if there's a way to get the same functionality without using i2cDevice. Same goes for other functions that fall in this category.

> I don't have a Windows machine to build it. I'll see if I can arrange one.

That would be great. Let me know if you have any more questions.

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,313 @@
+        wprintf(L"%d bytes partially transferred\n", bytesRead);
+        break;
+      case I2cTransferStatus::SlaveAddressNotAcknowledged:
+        wprintf(L"Slave address was not acknowledged\n");
```
Should there be a `return Nan::ThrowError(...)` after this `wprintf` call and possibly all other `wprintf` calls?

fivdi commented:
I'm unfamiliar with Windows IoT and this is likely to be visible in some of the comments and questions this time. Sorry about that.

On Linux, the i2c-bus methods can be regarded as atomic transactions. This makes it possible to write code that calls multiple async operations concurrently and the operations will not interfere with each other. An example of this can be seen in https://github.com/fivdi/i2c-bus/blob/master/example/two-devices.js. This example accesses two different devices at the same time asynchronously. In the background the operating system will ensure that the two different devices are actually accessed sequentially. I'd assume that the same is possible with Windows. Is this assumption correct?

Some of the C++ code uses try/catch blocks for exception handling, for example, [here](https://github.com/munyirik/i2c-bus/blob/win/src/win/i2c.cc#L138-L158). There are also sections of code without exception handling, for example, [here](https://github.com/munyirik/i2c-bus/blob/win/src/win/i2c.cc#L67-L68). Does the lack of try/catch blocks here mean that no exceptions can be thrown?

It looks like it will take a while before I'm in possession of a Windows 10 machine which is required for Windows IoT development if I'm not mistaken. On the other hand, the PR is non-intrusive and doesn't break any existing code. This means the PR could be merged without me actually knowing whether or not it functioned correctly.

If the PR is merged do you intend maintaining the merged code in the future or is this something you don't want to or would prefer not to do?

munyirik commented:
I believe your assumption is correct about the sequential access but let me confirm that and get back you you.
As for the try/catch blocks, they are actually not needed based on the documentation and examples I've look at so that was overkill. I'll take them out.
I do plan on maintaining the merged code at least to fix any issues in the existing code and also to complete the async functionality.

munyirik commented:
@fivdi i2cDevice methods are also atomic. The driver does all the necessary synchronization.

munyirik commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,313 @@
+        wprintf(L"%d bytes partially transferred\n", bytesWritten);
+        break;
+      case I2cTransferStatus::SlaveAddressNotAcknowledged:
+        wprintf(L"Slave address was not acknowledged\n");
```
after taking out the try blocks we should return failure if address isn't acknowledged or there is an invalid status value.

fivdi replied:
Yes.

For example, running the following on Linux will return failure if the device at address 0x48 is not on the I2C bus:
```js
var i2c = require('../'),
  i2c1 = i2c.openSync(1);

var DS1621_ADDR = 0x48,
  CMD_ACCESS_CONFIG = 0xac;

i2c1.writeByteSync(DS1621_ADDR, CMD_ACCESS_CONFIG, 0x01);
```
There error is detected and thrown [here](https://github.com/fivdi/i2c-bus/blob/v1.0.3/src/writebyte.cc#L68-L70), and this is what ends up on the screen:
```
/home/pi/i2c-bus/node_modules/i2c-bus/i2c-bus.js:249
  i2c.writeByteSync(peripheralSync(this, addr), cmd, byte);
      ^

Error: EIO, Input/output error
    at Error (native)
    at Bus.writeByteSync (/home/pi/i2c-bus/node_modules/i2c-bus/i2c-bus.js:249:7)
    at Object.<anonymous> (/home/pi/i2c-bus/node_modules/i2c-bus/example/error.js:7:6)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:456:32)
    at tryModuleLoad (module.js:415:12)
    at Function.Module._load (module.js:407:3)
    at Function.Module.runMain (module.js:575:10)
    at startup (node.js:160:18)
```

fivdi commented:
@munyirik excellent :)

munyirik commented:
I added the async versions of the initial functions I added. I need to do a little more testing before you merge.
Also if you get a chance to try it out that would ease my mind as well:-).

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,268 @@
+    DeviceInformationCollection^ i2cDeviceControllers = create_task(DeviceInformation::FindAllAsync(deviceSelector)).get();
+    
+    if (nullptr == i2cDeviceControllers) {
+      return Nan::ThrowError(Nan::ErrnoException(EPERM, "getController", "DeviceInformation::FindAllAsync failed to return controller(s)"));
```
This isn't a good idea. The Execute method is not called on the event loop thread (the main thread.) It's called on a worker thread so it's important steer clear of everything related to NAN, uv, Node, and V8. The exception to this rule is uv_async_send().

The best thing to do at the moment is to replace `return Nan::ThrowError(...);` with
```cpp
SetErrorNo(EPERM);
SetErrorSyscall("getController");
return;
```
Unfortunately the nice error message gets lost. [I2cAsyncWorker](https://github.com/fivdi/i2c-bus/blob/master/src/util.h#L4-L64) could be extended to support the nice error message.

munyirik replied:
Ah yeah I missed that. I ran into this problem in another place but didn't fix in all the files. Will fix.

fivdi commented on src/win/i2c.cc:
```diff
@@ -0,0 +1,268 @@
+    void Execute() {
+      I2cDevice^ device = create_task(I2cDevice::FromIdAsync(obj->GetControllerId(), settings)).get();
+      if (nullptr == device) {
+        return Nan::ThrowError(Nan::ErrnoException(EPERM, "createDevice",
```
This isn't a good idea. The Execute method is not called on the event loop thread (the main thread.) It's called on a worker thread so it's important steer clear of everything related to NAN, uv, Node, and V8. The exception to this rule is uv_async_send().

The best thing to do at the moment is to replace `return Nan::ThrowError(...);` with
```cpp
SetErrorNo(EPERM);
SetErrorSyscall("createDevice");
return;
```
Unfortunately the nice error message gets lost. [I2cAsyncWorker](https://github.com/fivdi/i2c-bus/blob/master/src/util.h#L4-L64) could be extended to support the nice error message.

munyirik replied:
Will fix.

munyirik commented on example/MCP9808-async.js:
```diff
@@ -0,0 +1,131 @@
+    function (cb) {
+      console.log("wake up MCP9808.... ");
+      shutdown_wake(0, function () {
+        cb(null);
```
tab

fivdi commented on win-i2c-bus.js:
```diff
@@ -0,0 +1,217 @@
+
+      bus._peripherals[addr] = device;
+
+      device.createDevice(addr, 1, 0, function (err) {
```
I see that busSpeed and sharingMode are moving closer to the API surface :)

fivdi commented:
> I need to do a little more testing before you merge.

Yes, please do this. Have you tested with more than one device on the bus at the same time? If not, can you do this also?

> Also if you get a chance to try it out that would ease my mind as well:-).

I want to, but I don't have the required hardware at the moment. The way the PR has been structured, it can be merged without breaking anything that's already there. If you have no objections, this is what I'll do. Ok?

munyirik commented:
Sounds good! you can merge it now. I'll test using more than one device next week and let you know the results.

fivdi commented:
@munyirik do you need a version published on npm?

fivdi commented:
@munyirik i2c-bus@1.1.0 was just published on npm

munyirik commented:
@fivdi thanks!

munyirik commented:
@fivdi FYI. I haven't got a second device to test with yet but I will in the next couple of days. I'll keep you posted.

munyirik commented:
@fivdi I should have a Windows version of example/two-devices.js in the next couple of days (sorry didn't get a chance to it this week).

fivdi commented:
@munyirik ok, no problem.

munyirik commented:
Sorry for the delay. I just got around to testing with two devices today and I didn't see any issues. I'll add the code for this as well as an example (synchrounous) for the Adafruit SI1145 UV sensor. PR coming shortly...
